File: app/packages/workbench/src/data/execution/execution.ts
import { FileType, Uri } from "vscode";
import type { FileSystemNode, Folder } from "../filesystem/file";
import { markRaw } from "vue";

/**
 * Represents the state of an execution.
 */
export type ExecutionState = "submitted" | "initializing" | "running" | "completed" | "cancelled" | "failed";

/**
 * Represents a file entry in an execution's result tree (from backend/plugin - flat structure).
 */
export interface ExecutionFileEntry {
    /**
     * Name of the file or directory (can be full path like "results/summary.m")
     */
    name: string;
    /**
     * Type of the entry (1 = file, 2 = directory)
     */
    type: number;
}

/**
 * File type constants matching backend FileType.
 */
export const ExecutionFileType = {
    FILE: 1,
    DIRECTORY: 2
} as const;

/**
 * Represents an execution.
 */
export interface Execution {
    /**
     * Unique identifier for the execution
     */
    id: string;
    /**
     * ID of the project this execution belongs to
     */
    projectId: string;
    /**
     * Path to the source file at the time of execution
     */
    filePath: string;
    /**
     * ID of the language plugin handling this execution
     */
    languageId: string;
    /**
     * Display name of the execution
     */
    name: string;
    /**
     * Current state of the execution
     */
    state: ExecutionState;
    /**
     * Optional progress indication text
     */
    progressText: string | null;
    /**
     * ISO 8601 timestamp when the execution was created/submitted
     */
    createdAt: string;
    /**
     * ISO 8601 timestamp when the execution started running, or null if not yet started
     */
    startedAt: string | null;
    /**
     * ISO 8601 timestamp when the execution finished, or null if not yet finished
     */
    finishedAt: string | null;
}

/**
 * Represents an execution with its result file tree.
 */
export interface ExecutionWithTree {
    /**
     * The execution metadata
     */
    execution: Execution;
    /**
     * Optional file tree for execution results (only available when completed)
     */
    fileTree: ExecutionFileEntry[] | null;
}

/**
 * Request payload for creating a new execution.
 */
export interface CreateExecutionRequest {
    /**
     * Path to the file to execute
     */
    filePath: string;
    /**
     * Arbitrary JSON data for the execution (e.g., function to execute)
     */
    data: unknown;
}

/**
 * Checks if an execution is in a terminal state (completed, cancelled, or failed).
 *
 * @param state The execution state to check
 * @returns True if the execution is in a terminal state
 */
export function isTerminalState(state: ExecutionState): boolean {
    return state === "completed" || state === "cancelled" || state === "failed";
}

/**
 * Checks if an execution can be cancelled.
 *
 * @param state The execution state to check
 * @returns True if the execution can be cancelled
 */
export function canCancelExecution(state: ExecutionState): boolean {
    return !isTerminalState(state);
}

/**
 * Converts a flat list of execution file entries into a tree structure.
 *
 * @param entries Flat list of file entries from backend/plugin
 * @param executionId Execution ID for creating unique node IDs
 * @param projectId Project ID for creating unique node URIs
 * @returns Root folder containing the tree structure
 */
export function buildExecutionFileTree(entries: ExecutionFileEntry[], executionId: string, projectId: string): Folder {
    const id = `/${projectId}/executions/${executionId}/files`;
    const root: Folder = {
        id,
        uri: markRaw(Uri.file(id)),
        name: "",
        type: FileType.Directory,
        parent: null,
        children: []
    };

    const sortedEntries = [...entries].sort((a, b) => {
        const aDepth = a.name.split("/").length;
        const bDepth = b.name.split("/").length;
        return aDepth - bDepth;
    });

    const folderMap = new Map<string, Folder>();
    folderMap.set("", root);

    for (const entry of sortedEntries) {
        const path = entry.name;
        const segments = path.split("/").filter((s) => s.length > 0);
        const name = segments[segments.length - 1] || "";
        const parentPath = segments.slice(0, -1).join("/");

        let parent = folderMap.get(parentPath);
        if (!parent) {
            parent = root;
        }

        const existingChild = parent.children.find((c) => c.name === name);
        if (existingChild) {
            continue;
        }

        if (entry.type === ExecutionFileType.DIRECTORY) {
            const id = `/${projectId}/executions/${executionId}/files/${path}`;
            const folder: Folder = {
                id,
                uri: markRaw(Uri.file(id)),
                name,
                type: FileType.Directory,
                parent,
                children: []
            };
            parent.children.push(folder);
            folderMap.set(path, folder);
        } else {
            const lastDotIndex = name.lastIndexOf(".");
            const extension = lastDotIndex > 0 ? name.substring(lastDotIndex) : "";
            const id = `/${projectId}/executions/${executionId}/files/${path}`;
            const file: FileSystemNode = {
                id,
                uri: markRaw(Uri.file(id)),
                name,
                type: FileType.File,
                parent,
                extension
            };
            parent.children.push(file);
        }
    }

    return root;
}

File: platform/common/src/main/kotlin/com/mdeo/common/model/Execution.kt
package com.mdeo.common.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Represents an execution in the system.
 *
 * @property id Unique identifier for the execution
 * @property projectId ID of the project this execution belongs to
 * @property filePath Path to the source file at the time of execution
 * @property languageId ID of the language plugin handling this execution
 * @property name Display name of the execution
 * @property state Current state of the execution
 * @property progressText Optional progress indication text
 * @property createdAt ISO 8601 timestamp when the execution was created/submitted
 * @property startedAt ISO 8601 timestamp when the execution started running, or null if not yet started
 * @property finishedAt ISO 8601 timestamp when the execution finished, or null if not yet finished
 */
@Serializable
data class Execution(
    val id: String,
    val projectId: String,
    val filePath: String,
    val languageId: String,
    val name: String,
    val state: String,
    val progressText: String?,
    val createdAt: String,
    val startedAt: String? = null,
    val finishedAt: String? = null
)

/**
 * Represents an execution with its result file tree.
 *
 * @property execution The execution metadata
 * @property fileTree Optional file tree for execution results (only available when completed)
 */
@Serializable
data class ExecutionWithTree(
    val execution: Execution,
    val fileTree: List<FileEntry>? = null
)

/**
 * Request payload for creating a new execution.
 *
 * @property filePath Path to the file to execute
 * @property data Arbitrary JSON data for the execution (e.g., function to execute)
 */
@Serializable
data class CreateExecutionRequest(
    val filePath: String,
    val data: JsonElement
)

/**
 * Response from plugin when creating an execution.
 *
 * @property name The display name for the execution (provided by the plugin)
 */
@Serializable
data class CreateExecutionResponse(
    val name: String
)

/**
 * Request payload for updating execution state (used by services via JWT).
 *
 * @property state New state of the execution
 * @property progressText Optional progress text
 */
@Serializable
data class UpdateExecutionStateRequest(
    val state: String,
    val progressText: String? = null
)

/**
 * Request payload forwarded to plugin for creating an execution.
 *
 * @property executionId Unique identifier for the new execution
 * @property project Project ID
 * @property filePath Path to the file being executed
 * @property data Arbitrary JSON data for the execution
 */
@Serializable
data class PluginCreateExecutionRequest(
    val executionId: String,
    val project: String,
    val filePath: String,
    val fileContent: String,
    val fileVersion: Int,
    val data: JsonElement,
    val contributionPlugins: List<JsonElement> = emptyList()
)

/**
 * Response from plugin execution file tree endpoint.
 *
 * @property files List of file entries in the execution result
 */
@Serializable
data class ExecutionFileTreeResponse(
    val files: List<FileEntry>
)

/**
 * Response from plugin execution summary endpoint.
 *
 * @property summary The summary content for the execution result
 */
@Serializable
data class ExecutionSummaryResponse(
    val summary: String
)

/**
 * Execution state constants used across services.
 * Centralized here to avoid inconsistent definitions in multiple modules.
 */
object ExecutionState {
    const val SUBMITTED = "submitted"
    const val INITIALIZING = "initializing"
    const val RUNNING = "running"
    const val COMPLETED = "completed"
    const val CANCELLED = "cancelled"
    const val FAILED = "failed"
}
